src/dragonball/src/address_space_manager.rs

```diff
@@ -0,0 +1,865 @@
+    FileInvalid(String),
+
+    /// Memory file provided is invalid due to empty memory type
+    #[error("memory type provided to address manager {0} is invalid")]
+    TypeInvalid(String),
+
+    /// Failed to set size for memory file.
+    #[error("address manager failed to set size for memory file")]
+    SetFileSize(#[source] std::io::Error),
+
+    /// Failed to unlink memory file.
+    #[error("address manager failed to unlink memory file")]
+    UnlinkFile(#[source] nix::Error),
+
+    /// Failed to duplicate fd of memory file.
+    #[error("address manager failed to duplicate memory file descriptor")]
+    DupFd(#[source] nix::Error),
+
+    /// Failure in accessing the memory located at some address.
+    #[error("address manager failed to access guest memory located at 0x{0:x}")]
+    AccessGuestMemory(u64, #[source] vm_memory::mmap::Error),
+
+    /// Failed to create GuestMemory
+    #[error("address manager failed to create guest memory object")]
+    CreateGuestMemory(#[source] vm_memory::Error),
+
+    /// Failure in initializing guest memory.
+    #[error("address manager failed to initialize guest memory")]
+    GuestMemoryNotInitialized,
+
+    /// Failed to mmap() guest memory
+    #[error("address manager failed to mmap() guest memory into current process")]
+    MmapGuestMemory(#[source] vm_memory::mmap::MmapRegionError),
+
+    /// Failed to set KVM memory slot.
+    #[error("address manager failed to configure KVM memory slot")]
+    KvmSetMemorySlot(#[source] kvm_ioctls::Error),
+
+    /// Failed to set madvise on AddressSpaceRegion
+    #[error("address manager failed to set madvice() on guest memory region")]
+    Madvise(#[source] nix::Error),
+
+    /// join threads fail
+    #[error("address manager failed to join threads")]
+    JoinFail,
+
+    /// Failed to create Address Space Region
+    #[error("address manager failed to create Address Space Region {0}")]
+    CreateAddressSpaceRegion(#[source] AddressSpaceError),
+}
+
+type Result<T> = std::result::Result<T, AddressManagerError>;
+
+/// Parameters to configure address space creation operations.
+pub struct AddressSpaceMgrBuilder<'a> {
+    mem_type: &'a str,
+    mem_file: &'a str,
+    mem_index: u32,
+    mem_suffix: bool,
+    mem_prealloc: bool,
+    dirty_page_logging: bool,
+    vmfd: Option<Arc<VmFd>>,
+}
+
+impl<'a> AddressSpaceMgrBuilder<'a> {
+    /// Create a new [`AddressSpaceMgrBuilder`] object.
+    pub fn new(mem_type: &'a str, mem_file: &'a str) -> Result<Self> {
+        if mem_type.is_empty() {
+            return Err(AddressManagerError::TypeInvalid(mem_type.to_string()));
+        }
+        Ok(AddressSpaceMgrBuilder {
+            mem_type,
+            mem_file,
+            mem_index: 0,
+            mem_suffix: true,
+            mem_prealloc: false,
+            dirty_page_logging: false,
+            vmfd: None,
+        })
+    }
+
+    /// Enable/disable adding numbered suffix to memory file path.
```

Comment:
It unclear from this comment why you might want to toggle this? What is this feature used for?

Reply:
Explanation added.